### DISCUSSION TITLE
Modbus TCP server

### DISCUSSION THREAD

**Discussion by KepLer100500** [apache/plc4x#1858] [Q&A] [answered]
Hello, colleagues. Could you tell me if it's possible to implement a Modbus emulator server using PLC4J?

**Comment by chrisdutz** [comment, marked as answer]
Well technically all the datatypes are there with which you could do that, but you would still need to implement the server logic, because that doesn't exist.

**Comment by glcj** [comment]
Hi,

Currently in the tools section we are implementing a tool that includes the Modbus server, here is the link.

https://github.com/apache/plc4x-extras/tree/feature/merlot

Some features,

. Multiple devices.
. Configurable server port.
. Dynamically configured by command line, or statically by configuration files.
. Flexible addressing area (defined by the user).
. Can be run as a service.
. Signal simulator.
. Integrated in OSGi environment.
. Battle tested.

Points for improvement
. Works only as holding registers.
. Duments

If you are interested in trying it, I will support you with any questions.

Kind regards.
